### PR TITLE
Feat  new locking

### PR DESCRIPTION
**Pull request by Blu-J**
## Test 

- [x] Config
- [x] Context
- [x] Control
- [x] Dependencies
- [x] fsck?
- [x] disk init
- [x] install cleanup
- [x] install update
- [x] manager health
- [x] Util Logger
- [x] Version?

## Test Install/ Uninstall

Videos showing a removal and install of service

https://user-images.githubusercontent.com/2364004/164785705-210c1bb3-8bfb-4a4b-b112-e3e162ffdcb9.mp4

https://user-images.githubusercontent.com/2364004/164785712-43b14e47-c4c1-4b97-8e28-7239a7c6e857.mp4

## Test - Config

https://user-images.githubusercontent.com/2364004/164756599-24730c8f-14ef-41f2-85aa-dbd079bc2948.mp4

## Test - Logs

Snap of the system logs
image
Snap of an app logs
image

## Test - Version

Snap of version
image

### CODE REVIEW

File: backend/src/config/mod.rs [hunk 1]
```diff
@@ -229,29 +416,23 @@ pub async fn set_dry(
...
+            &timeout,
+            true,
+            &mut BTreeMap::new(),
+            &mut breakages,
+            &locks,
+        )
         .await?;
+    }
+
+    locks.configured.set(&mut tx, true, id.to_string()).await?;
```
Comment (Blu-J) [inline backend/src/config/mod.rs:435]: Still haven't figured out how to not convert to string for this. Originally I was playing with the 'static lifetime but that wouldn't work...

File: backend/src/config/mod.rs [hunk 2]
```diff
@@ -568,3 +733,7 @@ pub fn configure_rec<'a, Db: DbHandle>(
     }
     .boxed()
 }
+#[instrument]
+pub fn not_found() -> Error {
```
Comment (Blu-J) [inline backend/src/config/mod.rs:737]: Should this be part of the db things? I don't know where this should live, or if it should.

File: backend/src/dependencies.rs [hunk 3]
```diff
@@ -660,21 +665,19 @@ pub async fn add_dependent_to_current_dependents_lists<
...
     current_dependencies: I,
+    locks: &dyn CurrentDependentsLock,
 ) -> Result<(), Error> {
     for (dependency, dep_info) in current_dependencies {
-        if let Some(dependency_model) = crate::db::DatabaseModel::new()
-            .package_data()
-            .idx_model(&dependency)
-            .and_then(|pkg| pkg.installed())
-            .check(db)
+        let mut dependency_dependents = locks
```
Comment (Blu-J) [inline backend/src/dependencies.rs:671]: Removed the mutation model for now, I like this more but the real reason was for simplicity and POC

File: backend/src/config/mod.rs [hunk 4]
```diff
@@ -215,6 +222,177 @@ pub fn set(
...
 }
 
+/// So, the new locking finds all the possible locks and lifts them up into a bundle of locks.
+/// Then this bundle will be passed down into the functions that will need to touch the db, and
+/// instead of doing the locks down in the system, we have already done the locks and can
+/// do the operation on the db.
+/// An UnlockedLock has two types, the type of setting and getting from the db, and the second type
+/// is the keys that we need to insert on getting/setting because we have included wild cards into the paths.
+pub struct ConfigLocks {
+    configured: UnlockedLock<bool, String>,
```
Comment (Blu-J) [inline backend/src/config/mod.rs:232]: Changed to verified locks

File: backend/src/config/mod.rs [hunk 5]
```diff
@@ -215,6 +222,177 @@ pub fn set(
...
+            .package_data()
+            .star()
+            .installed()
+            .map(|x| x.status())
+            .make_locker(LockType::Write)
+            .add_to_keys(&mut locks);
+
+        let skeleton_key = db.lock_all(locks).await?;
+        Ok(Self {
+            configured: configured.unlock(&skeleton_key)?,
```
Comment (Blu-J) [inline backend/src/config/mod.rs:350]: verified

**Comment by ProofOfKeags** [comment]
Reviewed in person.

**Comment by chrisguida** [comment]
Why is the OOMkill prevention commit in here? Wasn't #1402 merged a while back?

**Comment by ProofOfKeags** [comment]
Merging this without the patchdb change being merged seems sus...

**Comment by dr-bonez** [comment]
I'm not sure how this got merged - I didn't even look at this today...

**Comment by dr-bonez** [comment]
Ok, so apparently if you force push a branch that has already been merged to a branch that has an open pr, github thinks that it has already been merged.

**Comment by Blu-J** [comment]
Something weird happened; I think I pushed a older version of master up, so this pr was a no-op, and got auto merged or had no protections?
Moved to pr https://github.com/Start9Labs/embassy-os/pull/1422

**Pull request merged**